### DISCUSSION TITLE
How to create zip with folders?

### DISCUSSION THREAD

**Discussion by galvakojis** [gildas-lormeau/zip.js#402] [Q&A]
Can I export zip file which contains multiple folders with multiple files?

for example:

- A folder
    - --a file 1
    - --a file 2
    - --a file 3
- B folder
    - --b file 1
- C folder
    - --c file 1
    - --c file 2
    
    I am not sure i this is correct, and how later add files to 'A folder'?
```

      const zipFileWriter = new TransformStream();

      const zipFileBlobPromise = new Response(zipFileWriter.readable).blob();

      const zipWriter = new ZipWriter(zipFileWriter);

      let ls = await zipWriter.add('A folder', undefined, {directory: true, });
```

**Comment by gildas-lormeau** [comment]
You just have to include the folder in the filename. Actually, folders are not really useful in a zip file. You can simply avoid creating them explicitly and it should work fine. I confirm the code you posted seems correct though.

**Reply by galvakojis** [reply]
> You just have to include the folder in the filename. Actually, folders are not really useful in a zip file. You can simply avoid creating them explicitly and it should work fine.

it is the requirement for my application, data needs to have structure inside the zip file

**Reply by gildas-lormeau** [reply]
I just pushed an example of code showing how to handle directories, see https://github.com/gildas-lormeau/zip.js/blob/897282471593fe917f07323d16ea4078c332b0f0/tests/all/test-directory.js

**Reply by galvakojis** [reply]
> You just have to include the folder in the filename. Actually, folders are not really useful in a zip file. You can simply avoid creating them explicitly and it should work fine. I confirm the code you posted seems correct though.

thank you for help, but I already achieved a result a bit different, is it good?


```


    const zipFs = new fs.FS();
    await Promise.all(links.map( async (oneOfLink: Links, index) => {

      const directoryName: string = (oneOfLink.name) ?  oneOfLink.name : 'undefined';

      const entry = await zipFs.addDirectory(`${directoryName}`);

        const pdf =  new Blob([arraybuffer], {type: mimeType});
        if(pdf?.blob){
          await entry.addBlob(`${directoryName}.pdf`, pdf.blob)
      }
    }));
      const blob = await zipFs.exportBlob({ relativePath: true, bufferedWrite: true });

      const anchorElement = document.createElement('a');
      anchorElement.href = URL.createObjectURL(blob);
      anchorElement.download = `new.zip`;
      anchorElement.click();

```

And I get what I expect:

One zip file, inside two Folders in each folder one file.

**Reply by gildas-lormeau** [reply]
Yes, you can also achieve this with the `FS` API as you did.

**Reply by Nedgeva** [reply]
Found interesting case - zip archive with folder structure like
`folder-name.yml/subfolder/.../another-one/file.some`

So this structure created w/ zip.js perfectly fine and archive itself is ok.
When I'm unpacking generated zip w/ built-in zip manager in Windows 10 explorer it's still ok.
But when i'm using 7zip (latest available binary) it's complains that files in `folder-name.yml` folder can't be extracted and written to disk, however inner structure e.g. `subfolder/.../another-one/file.some` extracted ok. I guess there something unpacker and fs related, ins't it?
Same structure w/ absolutely same file contents packed to zip in 7zip gets extracted w/o any issues at all.

I was using simple ` zipWriter.add(fileName, new TextReader(file.contents))` approach, no zipFS involved.

**Reply by gildas-lormeau** [reply]
Yes, it's probably a weird limitation in 7zip. In zip.js, all names are valid because the library has not this responsibility.
